File: base/src/main/java/compass_system/storagecores/base/data/tiers/TiersFile.java
package compass_system.storagecores.base.data.tiers;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import compass_system.storagecores.base.Constants;
import compass_system.storagecores.base.data.ReplacingMap;
import net.minecraft.resources.ResourceLocation;

import java.util.Map;

public record TiersFile(Map<ResourceLocation, TierEntry> entries, boolean replace) implements ReplacingMap<TierEntry> {
    public static final Codec<TiersFile> CODEC = RecordCodecBuilder.create(
            instance -> instance.group(
                            Codec.unboundedMap(Constants.normalisedResourceLocationCodec(), TierEntry.CODEC).fieldOf("values").forGetter(TiersFile::entries),
                            Codec.BOOL.optionalFieldOf("replace", Boolean.FALSE).forGetter(TiersFile::replace)
                    )
                    .apply(instance, TiersFile::new)
    );

    @Override
    public boolean shouldReplace() {
        return replace;
    }

    @Override
    public Map<ResourceLocation, TierEntry> values() {
        return entries;
    }
}

File: base/src/main/kotlin/compass_system/storagecores/base/data/tiers/TiersFile.kt
package compass_system.storagecores.base.data.tiers

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import compass_system.storagecores.base.Constants
import compass_system.storagecores.base.data.ReplacingMap
import net.minecraft.resources.ResourceLocation

data class TiersFile(val values: Map<ResourceLocation, TierEntry>, val replace: Boolean) : ReplacingMap<TierEntry> {
    override fun replaceExistingEntries() = replace

    override fun entries() = values

    companion object {
        val codec : Codec<TiersFile> = RecordCodecBuilder.create { instance ->
            instance.group(
                Codec.unboundedMap(Constants.normalisedResourceLocationCodec(), TierEntry.codec).fieldOf("values").forGetter(TiersFile::values),
                Codec.BOOL.optionalFieldOf("replace", false).forGetter(TiersFile::replace)
            ).apply(instance, ::TiersFile)
        }
    }
}
